analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs

```diff
@@ -198,7 +213,7 @@ private IEnumerable<string> FindCredentialWords(string variableName, string vari
                 return Enumerable.Empty<string>();
             }
 
-            var match = passwordValuePattern.Match(variableValue);
+            var match = PasswordValueRegex(CredentialWords).Match(variableValue);
```

Comment:
I don't understand the `PasswordValueRegex` method, does `CredentialWords` can change within the same analysis and between files or nodes? (Never implemented a parametrized rule yet)
In other words, why `PasswordValuePattern` is a Dictionary rather than a Regex?

As I understand it, `PasswordValuePattern` should always have only one entry.

Reply:
Most likely yes, but I'm not totally sure. Accessing the regex via the `CredentialWords` key seemed the safest way to do it. The `CredentialWords` can change at any time now and we still do the right thing.

Reply:
For the record: The way parametrized rules work is conceptually wrong. Properties should NOT be stored on the class. But they are, and it wasn't a big problem so far (this might change in the future with dedicated QP for tests for example). As of now, this isn't a problem.

The ConcurrentDictionary is an overkill in this case. All that is needed is to store a new regex in a single field directly. 

In practice:
VS and NuGet analysis have parametrized rules turned off.
S4NET analysis loads parameters in CompilationStart for each compilation. If we store string, or compiled regex is equivalent. Storing regex in a dictionary is overkill here, as there is nothing to win - store of nonchanged value should not regenerate the regex as well

Reply:
This sounds very brittle to me. I use a `Tuple<string, Regex>` so we are at least sure we return the right regex. Taking a strong dependency on the current situation might cause an "error in the distance" later.

Reply:
Event the tuple is not needed. The "desired" regex (string) changes only on the property change. And that should rebuild currently used regex.

Once we fix the design issue later, we'll need to review and rework all usages. And it will go in direction of fixing all of them automatically. So the Tuple would only add confusion.

Reply:
I don't want to appear in the git history for such a change. I turned it back into an instance field and removed the compiled flag.